### PR TITLE
feat(#1357): Lexical chunk retriever

### PR DESCRIPTION
**Pull request by patelchaitany**
## Description

I Implemented Lexical Chunk Retriever In the LexicalRetriever class is Inherite the BaseRetriever and The DocumentChunk are lazy loaded when first time query is made because it save time during object initialization
and the function get_context and the get_completion are Implemented same as the ChunksRetriever the only diffrence is that the DocumentChunk are converted to match the output type of the ChunksRetriever using function get_own_properties in the utils.

## Type of Change

- [-] Bug fix (non-breaking change that fixes an issue)
- [-] New feature (non-breaking change that adds functionality)
- [-] Breaking change (fix or feature that would cause existing functionality to change)
- [-] Documentation update
- [-] Code refactoring
- [-] Performance improvement
- [-] Other (please specify):

## Changes Made

     - Added LexicalRetriever base class with customizable tokenizer & scorer
     - Implemented caching of DocumentChunk tokens and payloads 
     - Added robust initialization with error handling and logging - Implemented get_context with top_k ranking and optional scores 
     - Implemented get_completion consistent with BaseRetriever interface 
     - Added JaccardChunksRetriever demo using set/multiset Jaccard similarity 
     - Support for stopwords and multiset frequency-aware similarity - Integrated logging for initialization, scoring, and retrieval

## Testing

    - Manual tests: initialized retriever, retrieved chunks with toy corpus 
    - Edge cases: empty corpus, empty query, scorer/tokenizer errors 
    - Verified Jaccard similarity results for single/multiset cases 
    - Code formatted and linted

## Pre-submission Checklist

- [-] **I have tested my changes thoroughly before submitting this PR**
- [-] **This PR contains minimal changes necessary to address the issue/feature**
- [-] My code follows the project's coding standards and style guidelines
- [-] I have added tests that prove my fix is effective or that my feature works
- [-] I have added necessary documentation (if applicable)
- [-] All new and existing tests pass
- [-] I have searched existing PRs to ensure this change hasn't been submitted already
- [-] I have linked any relevant issues in the description
- [-] My commits have clear and descriptive messages

## Related Issues

Relates to  #1392
## Additional Notes

Int the cognee/modules/chunking/models/DocumentChunk.py
don't remove the optional  from is_part_of attributes.

## DCO Affirmation
I affirm that all code in every commit of this pull request conforms to the terms of the Topoteretes Developer Certificate of Origin.

### CODE REVIEW

**Comment by patelchaitany** [comment]
@hande-k can you pls review the changes.

**Comment by hande-k** [comment]
Hi @patelchaitany, thanks for your contribution! We will review this PR shortly. Meanwhile, could you take a look at the issues already raised by Code Rabbit and address them?

We appreciate your time and effort and looking forward to merge it to the codebase!

**Comment by patelchaitany** [comment]
resolved the some nitpick comment which can cause some bug.also added improvement suggested by the coderabbit.

**Comment by patelchaitany** [comment]
@hande-k this is status for the implementation.

## Implementation Status
- [x] Implemented `JaccardChunksRetriever` inheriting from `LexicalRetriever`.
- [x] Features available but **disabled**:
  - [x] Stop words filtering.
  - [x] Multiset Jaccard scoring (frequency aware).

## Non-Functional Requirements
- [x] Empty corpus → returns `[]`.
- [x] Empty/degenerate query → returns `[]`.
- [x] Tokenizer errors → logged, return `[]`.
- [x] Scorer errors → logged, fallback to `0.0`.
- [x] Non-numeric scores → treated as `0.0`.
- [x] Cache reused across multiple queries.
- [x] Cache refresh only by process restart.

## Logging
- [x] Candidate count logged.
- [x] Tokenize time logged.
- [x] Score time logged.
- [x] top_k returned logged.
- [x] Query length logged.

## Integration
- [x] Added `CHUNKS_LEXICAL` to `SearchType`.
- [x] Updated dispatcher to call `JaccardChunksRetriever`.
- [x] Verified no regression in existing retrievers.

## Acceptance Criteria
- [x] Returns same payload shape as `ChunksRetriever`.
- [x] Scores are raw, no tie-breaks.
- [x] Cache reused across multiple queries.
- [x] Query `"cat"` retrieves chunk with `"cat dog"`.
- [x] Empty corpus returns `[]`.
- [x] Empty/degenerate query returns `[]`.
- [x] Logs confirm scoring path and top_k size.
- [x] `CHUNKS_LEXICAL` type works end-to-end.

File: cognee/modules/retrieval/lexical_retriever.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+from cognee.modules.retrieval.exceptions.exceptions import NoDataError
+from heapq import nlargest
+logger = get_logger("LexicalRetriever")
+
+class LexicalRetriever(BaseRetriever):
+    _global_chunks: dict[str, Any] = {}
+    _global_payloads: dict[str, Any] = {}
+    _global_initialized = False
+    _global_init_lock = asyncio.Lock()
+
```
[the remark this answers is not in the document]
Comment (patelchaitany) [reply cognee/modules/retrieval/lexical_retriever.py]: using the instance scope is not good idea because when we do the search on the database every time it create new instance of the class and every time we have to retrieve data from the db so it takes time for processing and retrieving large db so i used class level caches which can be used across multiple query.

**Comment by patelchaitany** [comment]
@Vasilije1990 or @hande-k can you review the PR I solved all comment raised by code rabbit. And tested on locally.

**Comment by hande-k** [comment]
hey @patelchaitany, thanks for the updates! we'll look into it shortly

**Comment by lxobr** [comment]
Thanks again for this contribution! We really appreciate the time and thought you put into it. The core of your work (lexical_retrieval and jaccard_retrieval) is very close to what we had in mind, and that’s great to see! 🔥

I also noticed that you touched quite a few files beyond the new retrievers. Reading through, I can see the logic behind those changes and I appreciate the initiative. That said, with the bigger picture in mind, we’d like this PR to stay focused just on the new retrievers, plus the update to the search type enum. 

A bit of context: it’s perfectly fine not to expose with_scores directly in cognee.search yet, as advanced users who need it can import the retriever directly. If you feel there are typing improvements needed, please feel free to open a separate issue. There’s also no need to adjust whitespace in unrelated files. Also, the retriever should be implemented without relying on global context variables. Finally, the existing examples in the code base should not be modified with this PR.

With this in mind, I would like to kindly ask you to revert all the changes that touch the existing files in the codebase, except the search type enum. If the retrievers currently rely on those edits, please adapt their implementation so they work independently. Once the scope is narrowed back and the retrievers are self-contained, I’ll be happy to take an even deeper dive into the retrievers themselves, provide feedback if needed, and gladly merge. 🦾

A little heads up: for testing, I’ll run your retrievers in a modified version of `simple_example.py` (by swapping out the default text and queries, and search type/retriever). If you’d like, you can even add a new example and use `simple_example.py` as a starting template to showcase your retriever. 

Thanks again for contributing (and for reading the long comment 😅). I am glad that you reacted to the issue very quickly and I am really looking forward to reviewing the next iteration! 🚀  And who knows, maybe we can take it even further in a follow-up issue by implementing BM25.

**Comment by patelchaitany** [comment]
@lxobr Regarding the adopting this with implementing lexical retriever without global context introduce one problem when we do search for different dataset the cashing if we changes the dataset to different dataset then it end up using old dataset cashe

**Comment by lxobr** [comment]
@patelchaitany , thanks for the comment, and gllad you’re thinking about dataset handling so carefully. 
In practice this isn’t an issue though: when retrievers are called through the `search` interface, the user/dataset context is already taken care of. If you look at other retrievers (for example `ChunksRetriever`), you’ll see that none of them work directly on a dataset level, precisely because `search` already prepares everything for the retriever.  Hope this helps. 
Really appreciate that you’re diving deep into this! 🙌 Please keep me updated and let me know if you need any further support

**Comment by patelchaitany** [comment]
@lxobr now i have removed the white space and narrow down the scope for the changes also i made the LexicalRetrival work independently and I also removed the the with_score variable form search function. and the regarding typing change to the DocumentChunk.py i have to do becuse in the LexicalRetrival i am using from_dict function to do seralization form dict to object.

I will open new issue for the tryping change.

**Comment by lxobr** [comment]
Thanks for the updates and for narrowing the scope! This is already a big improvement, and I really appreciate the time and effort.  🚀 

About the typing change in `DocumentChunk`: that isn’t needed. It only feels necessary because you’re using `from_dict` to hydrate full objects inside the retriever. In practice, the retriever just needs `id` and `text` (or even just `text`), so the model can stay as-is.  So could you please revert those?

A simple way to unlock this is to use `get_filtered_graph_data` (instead of `get_graph_data`) with an attribute filter like  `{"type": ["DocumentChunk"]}`. That way you fetch only the chunks, avoid scanning the whole graph, and don’t need JSON repairs or model changes. It keeps the retriever lean and efficient.  

So I’d kindly ask you to drop the JSON fixing/checking and the `DocumentChunk` change. Once these couple of points are addressed, I’ll be happy to take a deeper dive into the retriever itself and get this ready to merge. Really appreciate the thoughtful work you’re putting in! 🙌

**Comment by patelchaitany** [comment]
#1437 while working on the current changes i have encountered the bug in the kuzu database adpter query so i have created issue for that bug and created pull request for that bug.can you review that bug @lxobr.
 
 
 the function get_filtered_graph_data other wise it is working with the other databases.

File: cognee/modules/retrieval/lexical_retriever.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        try:
+            query_tokens = self.tokenizer(query)
+        except Exception as e:
+            logger.error("Failed to tokenize query: %s", str(e))
+            return []
+
+        if not query_tokens:
+            logger.warning("Query produced no tokens")
+            return []
+
```
[the remark this answers is not in the document]
Comment (patelchaitany) [reply cognee/modules/retrieval/lexical_retriever.py:92]: it is tokenizer responsibility to give tokens in list format and tokenizer only need to return tokens as list nothing else.

**Comment by patelchaitany** [comment]
@lxobr is there any thing i need to change? I have removed all the unnecessary code and only using DocumentChunk now i you can take deeper dive in the lexical chunk class.

**Comment by lxobr** [comment]
Hey @patelchaitany, thanks for the updates and reaching out! I will review and test it today and will get back to you in a couple of hours, but I can say it looks quite good already. Appreciate your time and effort!

**Review by lxobr — DISMISSED** [review]
Thanks for the hard work! Really appreciate your effort and patience.
This looks good now, and happy to merge.
